### PR TITLE
Prepare to release 2.4.0beta1

### PR DESCRIPTION
**Pull request by ashb**
As I like to have the branch point for a relase as "late" as possible,
I'm committing the version change to main, and then branching off.

The release notes/changelog are _far_ from complete, but lets take what
newsfragments we have

### CODE REVIEW

**Review by ephraimbuddy — APPROVED** [review]

**Comment by ephraimbuddy** [comment]
Should we also change all references of `2.4.0.dev0` to `2.4.0beta1`? Mostly in docs/

**Comment by ashb** [comment]
```
CONTRIBUTORS_QUICK_START.rst:   Airflow source version: 2.4.0.dev0
docs/docker-stack/README.md:You can find the following images there (Assuming Airflow version `2.4.0.dev0`):
docs/docker-stack/README.md:* `apache/airflow:2.4.0.dev0` - the versioned Airflow image with default Python version (3.7 currently)
docs/docker-stack/README.md:* `apache/airflow:2.4.0.dev0-pythonX.Y` - the versioned Airflow image with specific Python version
docs/docker-stack/README.md:* `apache/airflow:slim-2.4.0.dev0`           - the versioned Airflow image with default Python version (3.7 currently)
docs/docker-stack/README.md:* `apache/airflow:slim-2.4.0.dev0-pythonX.Y` - the versioned Airflow image with specific Python version
docs/docker-stack/docker-examples/extending/add-apt-packages/Dockerfile:FROM apache/airflow:2.4.0.dev0
docs/docker-stack/docker-examples/extending/add-build-essential-extend/Dockerfile:FROM apache/airflow:2.4.0.dev0
docs/docker-stack/docker-examples/extending/add-providers/Dockerfile:FROM apache/airflow:2.4.0.dev0
docs/docker-stack/docker-examples/extending/add-pypi-packages/Dockerfile:FROM apache/airflow:2.4.0.dev0
docs/docker-stack/docker-examples/extending/custom-providers/Dockerfile:FROM apache/airflow:2.4.0.dev0
docs/docker-stack/docker-examples/extending/embedding-dags/Dockerfile:FROM apache/airflow:2.4.0.dev0
docs/docker-stack/docker-examples/extending/writable-directory/Dockerfile:FROM apache/airflow:2.4.0.dev0
docs/docker-stack/entrypoint.rst:  docker run -it apache/airflow:2.4.0.dev0-python3.6 bash -c "ls -la"
docs/docker-stack/entrypoint.rst:  > docker run -it apache/airflow:2.4.0.dev0-python3.6 python -c "print('test')"
docs/docker-stack/entrypoint.rst:   docker run -it apache/airflow:2.4.0.dev0-python3.6 airflow webserver
docs/docker-stack/entrypoint.rst:  > docker run -it apache/airflow:2.4.0.dev0-python3.6 help
docs/docker-stack/entrypoint.rst:    FROM airflow:2.4.0.dev0
docs/docker-stack/entrypoint.rst:    FROM airflow:2.4.0.dev0
docs/docker-stack/entrypoint.rst:      apache/airflow:2.4.0.dev0-python3.8 webserver
docs/docker-stack/entrypoint.rst:      apache/airflow:2.4.0.dev0-python3.8 webserver
docs/docker-stack/entrypoint.rst:      apache/airflow:2.4.0.dev0-python3.8 webserver
setup.py:version = '2.4.0.dev0'
```

I think those can stay until 2.4.0 is released GA?

File: RELEASE_NOTES.rst [hunk 1]
```diff
@@ -21,6 +21,127 @@
...
+          start_date=datetime(2021, 1, 1),
+          schedule=EventsTimetable(event_dates=[pendulum.datetime(2022, 4, 5)]),
+      ):
+          ...
+
+  (#25410)
+- Removal of experimental Smart Sensors
+
+  Smart Sensors were added in 2.0 and deprecated in favour of Deferable operators in 2.2, and have now been removed. (#25507)
+- The ``contrib`` packages have now dynamically generated modules and while users can continue using the deprecated contrib classes, they are no longer visible for static code check tools and will be reported as missing. It is recommended for the users to move to non-deprecated classes. (#26153)
```
Comment (potiuk) [inline RELEASE_NOTES.rst]:
```suggestion
- The ``airflow.contrib`` packages and deprecated modules from Airlfow 1.10 in ``airflow.hooks``, ``airflow.operators``, ``airflow.sensors`` packages, have now dynamically generated modules and while users can continue using the deprecated contrib classes, they are no longer visible for static code check tools and will be reported as missing. It is recommended for the users to move to non-deprecated classes. (#26153, #26179, #26167)
```

**Review by potiuk — COMMENTED** [review]
Small nit (forgot to update newsfragment)

**Review by potiuk — APPROVED** [review]

**Comment by uranusjr** [comment]
The auto register fragment is _so_ long I think we should change the items into subsections. Should this be done here or in a later PR?

**Comment by ashb** [comment]
Yeah, just noticed that, and fixing the check in version head map to understand that.

My "fix" was to just remove the `--version-suffix-for-pypi` from the build images workflow -- we dont' need it there do we?

**Comment by potiuk** [comment]
Fix here: https://github.com/apache/airflow/pull/26230

**Comment by potiuk** [comment]
> My "fix" was to just remove the `--version-suffix-for-pypi` from the build images workflow -- we dont' need it there do we?

We need that at least to indicate in the production imaage that the version  is not the "released" one. It might be confusing for someone who looks at the image for testing for example so I prefer to override it to dev0 always.

It's more needed for providers - becasue they will fail to generate documentation (this is a protection agains accidentally trying to release the same version again - so for providers adding suffix on CI serves also this role, but I think keeping dev0 always when it is not specified for the regular CI builds is a good thing.

**Comment by ashb** [comment]
> The auto register fragment is _so_ long I think we should change the items into subsections. Should this be done here or in a later PR?

Probably later PR -- we'll need to add a load more anyway (there's nothing in there about Datasets for instance)

File: tests/www/views/test_views_base.py [hunk 2]
```diff
@@ -35,16 +34,14 @@ def test_index(admin_client):
...
 
 
-def test_doc_urls(admin_client):
+def test_doc_urls(admin_client, monkeypatch):
+    # Mocking this way is tying the test closer to the implementation much more than I'd like. :shrug:
+    from airflow.www.views import AirflowBaseView
+
+    monkeypatch.setitem(AirflowBaseView.extra_args, 'get_docs_url', lambda _: "!!DOCS_URL!!")
     resp = admin_client.get('/', follow_redirects=True)
-    if "dev" in version.version:
```
Comment (ashb) [inline tests/www/views/test_views_base.py:40]: I did this, as otherwise we are just re-implementing the get_docs_url logic here in the test.

**Comment by ashb** [comment]
Looks like CI wants us to update the versions anyway, cool I'll do that then.

**Comment by potiuk** [comment]
> The auto register fragment is _so_ long I think we should change the items into subsections. Should this be done here or in a later PR?

Yeah. We should split it definitely (and can be done separately)

**Comment by potiuk** [comment]
> Looks like CI wants us to update the versions anyway, cool I'll do that then.

Yeah. Those are independent from `--suffix` the suffix is only for package preparation.

**Comment by potiuk** [comment]
docs failiung though

**Comment by ashb** [comment]
> Incorrect Spelling: 'favour'
> Line with Error: 'Smart Sensors were added in 2.0 and deprecated in favour of Deferable operators in 2.2, and have now been removed.'
> Line Number: 127

🫖

**Comment by potiuk** [comment]
BTW. I think the problem is that the newsfragments are not checked for spelling I will take a look at that

**Comment by ashb** [comment]
> BTW. I think the problem is that the newsfragments are not checked for spelling I will take a look at that

Yeah, I was thinking the same. A number of errors I had to fix.

**Comment by ashb** [comment]
Tested docs locally, living fast-n-loose :shipit:

**Pull request merged**